source/common/http/filter_manager.h

```diff
@@ -106,9 +127,35 @@ class HttpResponseHeadersDataInput : public HttpHeadersDataInputBase<ResponseHea
   }
 };
 
+class HttpResponseHeadersDataInputFactory : public Matcher::DataInputFactory<HttpMatchingData> {
+public:
+  std::string name() const override { return "response-headers"; }
+  Matcher::DataInputPtr<HttpMatchingData>
+  createDataInput(const Protobuf::Message& config,
+                  ProtobufMessage::ValidationVisitor& validation_visitor) override {
+    const auto& typed_config = MessageUtil::downcastAndValidate<
+        const envoy::type::matcher::v3::HttpRequestHeaderMatchInput&>(config, validation_visitor);
```

Comment:
```suggestion
        const envoy::type::matcher::v3::HttpResponseHeaderMatchInput&>(config, validation_visitor);
```

?

Maybe just template this as you are likely going to end up with trailer versions also?